File: plugin_files/bin/scripting/exports.lua
function export(exportName, callback)
    AddEventHandler("export:"..GetCurrentPluginName()..":"..exportName, function(event, ...)
        event:SetReturn(callback(...))
        return EventResult.Stop
    end)
end

exports = {}

setmetatable(exports, {
    __index = function(tbl, pluginName)
        local plugin = pluginName

        return setmetatable({}, {
            __index = function(tbl, key)
                return function(self, ...)
                    local eventRt, eventData = TriggerEvent("export:"..plugin..":"..key, ...)
                    if eventRt ~= EventResult.Stop then
                        print("Export '"..key.."' in '"..plugin.."' doesn't exists.")
                    end
                    return eventData:GetReturn()
                end
            end,

            __newindex = function(tbl, key, val) 
                print("You can't create a new index on this table.")
            end
        })
    end,

    __newindex = function(tbl, key, val) 
        print("You can't create a new index on this table.")
    end
})

File: plugin_files/bin/scripting/exports.js
function LoadExports(global) {
    let pluginProxyCache = {}

    global.exp = (exportName, callback) => {
        AddEventHandler(`export:${GetCurrentPluginName()}:${exportName}`, (event, ...args) => {
            event.SetReturn(callback.apply(global, args))
            return EventResult.Stop
        })
    }

    global.exports = new Proxy({}, {
        get: (tbl, pluginName) => {
            if(!pluginProxyCache[pluginName]) {
                pluginProxyCache[pluginName] = new Proxy({}, {
                    get: (tbl, funcName) => {
                        return (...args) => {
                            let [eventRt, eventData] = TriggerEvent(`export:${pluginName}:${funcName}`, ...args)
                            if (eventRt != EventResult.Stop) {
                                console.log(`Export '${funcName}' in '${pluginName}' doesn't exists.`)
                            }
                            return eventData.GetReturn()
                        }
                    },
                    set: (tbl, key, val) => {
                        console.log("You can't create a new value on this object.")
                        return false
                    }
                })
            }
            return pluginProxyCache[pluginName]
        },
        set: (tbl, key, val) => {
            console.log("You can't create a new value on this object.")
            return false
        }
    })
}

LoadExports(globalThis)